notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,285 @@
+   or so slower than the intiial solution described below.  I added a
+   `CkValue` type that was more or less isomorphic to the usual values
+   except for builtins, where the value includes a count of arguments
+   and lists of the type and term arguments it's received so far.
+   Introducing the new type also required quite a lot of typeclass
+   stuff to make it acceptable to the CAM.  See
+   https://github.com/input-output-hk/plutus/pull/2215 .
+
+   I think that what's going on here is that for partially-applied
+   buitins we're forced to have lists _somewhere_: this trick allows
+   us to move them out of the AST (where the typechecker has to deal
+   with them) and into the machines.
+
+ * What I've done for the time being [this is my original attempt, in
+   `kwxm/alternative-cek`] is to modify the CK machine to use a
+   variant of the method used in the original CEK machine.  Whenever
+   we see an application `[M N]` where `M` isn't a lambda, we call
+   `termAsPrimIterApp` to check if it's of the correct form, and if it
+   is we look up the the arity of the function and compare it with the
+   number of arguments; if they match, we hand the application off to
+   the CAM.  This is a bit more efficient than the earlier method
+   because we only call the CAM once, but it still requires repeated
+   calls to the inefficient `termAsPrimIterApp`, which we otherwise
+   don't need.
+
+ * We could modify the `builtin` constructor in the AST to contain a
+   list of arguments that it's received so far.  That's probably a
+   really bad idea though, since it'd only be there for the benefit of the
+   CK machine and it's taking us back towards the problems we had with
+   saturated builtins.  Let's not do that.
+ 
+
+I'm not sure what we want to do here.  How much do we care about the
+CK machine these days?  I think we originally had it because it was
+pretty simple and easily formalised, and fitted in nicely with the
+reduction semantics.  Maybe that's not so important now?
+
+Another issue is that in the specification we've been skating around
+exactly how builtin evaluation works, but we'll need to be more
+precise (and independent of the implementation language) if and when
+we do away with static builtins.  Maybe the machine implementation
+should be one that's easily specified.  An argument in favour of the
+CK machine was its simplicity, but maybe it's not possible to be very
+simple if we deal with builtins properly.
+
+Also, our primary evaluator will eventually be untyped, and it'll
+presumably be a CEK machine.  We'll surely still want a typed
+evaluator as well though, so maybe that should be the CEK machine as
+well.  It's maybe tempting to use the typed CEK machine and make it a
+bit more robust by using the `TypeArg | TermArg` thing mentioned above
+to count the arities: that'll be slightly less efficient, but we'll
+only need it for the typed evaluator, where efficiency isn't paramount.
+
+
+## Higher-rank builtins
```

Comment:
I initially was confused when you mentioned "higher-rank" builtins, because I always thought those were about types having a `forall` inside the type of an argument. Here's what I see on [Haskell wiki](https://wiki.haskell.org/Rank-N_types):

> Normal Haskell '98 types are considered Rank-1 types. A Haskell '98 type signature such as
>
> `a -> b -> a`
>
> implies that the type variables are universally quantified like so:
>
> `forall a b. a -> b -> a`
>
> `forall` can be floated out of the right-hand side of `->` if it appears there, so:
>
> `forall a. a -> (forall b. b -> a)`
>
> is also a Rank-1 type because it is equivalent to the previous signature. 
>
> However, a `forall` appearing within the left-hand side of `(->)` cannot be moved up, and therefore forms another level or rank. The type is labeled "Rank-N" where N is the number of foralls which are nested and cannot be merged with a previous one. For example:
>
> `(forall a. a -> a) -> (forall b. b -> b)`
> is a Rank-2 type because the latter `forall` can be moved to the start but the former one cannot. Therefore, there are two levels of universal quantification. 

So is it correct to talk about higher-rank builtins here?

Reply:
Yeah, now I'm just confused. What do we have then? Can we always move the quantifiers to the front in our case?

Reply:
Yeah, that does sound right. We're talking about cases where we've got type arguments in the middle, where they could be floated to the left. However for us these are not quite the same, because type arguments delay evaluation!

Reply:
> However for us these are not quite the same, because type arguments delay evaluation!

Whatever, I think "higher-rank types" is an established terminology that doesn't quite fit in our case, but not totally sure.

Reply:
Now I think that we do have higher-rank builtins whatever we do, because they can certainly accept abstractions as arguments (although maybe they can't do a lot with them), and that lets you put a quantifier to the left of as many arrows as you want in the tree representation of the type, which is what higher-rank types are about.

So yes, whatever we've been talking about, I don't think it is higher-rank types: sorry about that.  I think things are a bit confused because again we're talking about multi-argument functions and they're not something that's commonly discussed in type theory.